### ISSUE TITLE
Kernelmode debugging in windows 7 vm

### ISSUE THREAD

**Issue by Nordgaren** [cheat-engine/cheat-engine#2417]
Hello!

I am currently trying to use CE in a windows 7 VM to show the stack during a buffer overflow ROP attack in HEVD, but I cannot read that region of memory.

I was also looking into getting kernel mode debugging working, just in case it lead me to an answer to the above question, but it's greyed out in settings. I enabled it in the registry and restarted the VM, but it is still greyed out.

I found this issue, but no mention of which version of CE I should use, but it does sound like maybe an old version is the answer to my problems! <https://github.com/cheat-engine/cheat-engine/issues/1131>

This would be on the latest version of Cheat Engine (Downloaded today in the VM).

Images for reference.

Thank you for your time!

![image](https://user-images.githubusercontent.com/83100346/206439183-0bf1e323-602e-46b8-bec3-ce9252fbf59b.png)
![image](https://user-images.githubusercontent.com/83100346/206439222-0e327e2e-a553-4c10-ad10-2742ba4ffeff.png)

**Comment by cheat-engine** [comment]
Looks like KVA shadowing is still enabled even after editing the registry. 
What you could do is use regedit and set "Use Kernel Debugger" to true in ce's settings. And make sure that the target application runs explicitly with admin rights so there won't be a less privileged usermode pagetable base

**Comment by Nordgaren** [comment]
`HKLM\SYSTEM\CurrentControlSet\Control\Session Manager\Memory Management\FeatureSettingsOverride` is 3 and I set `Use Kernel Debugger` to 1 inside `HKEY_USERS\S-1-5-21-835841959-2849677911-2169483152-1001\Software\Cheat Engine`. I even tried disabling the DBVM debugger inside the registry. I think I confirmed this is the right registry entry by opening settings and seeing that the registry edits for Use Kernel Debugger and Use DBVM got set back to what they are in settings.

I can read the PE Header at `HEVD.sys` but `HEVD.sys + 0x865B4` is `??` still.

Could this have to do with being in a VM or because I have VirtualKD installed with signed drivers turned off?

I tried using Chrome as admin and tried using the CE process itself, as admin, as well.

Would it be worth trying making my own small program to run as admin that actually calls the driver with `DeviceIoControl`?

Thank you, again! As I said before, I am working on a video about Kernel exploits in Windows 7, and I want to show the stack being updated in real-time as our buffer overflows onto the stack pointer, and I think Cheat Engine would be a great way to show that!

**Comment by cheat-engine** [comment]
not being able to read the memory might be because the memory is paged out, and dbk64 does not read kernelmode memory if it detects it hasn't been paged in

**Comment by Nordgaren** [comment]
Is this what you meant? Sorry for the noob question. I don't know much about how paging works with the kernel, yet. :eyes:

On a side note, when I try to run python and use CreateFileA to get a handle to the HEVD driver, it returns -1 with GetLasetError() of 3. Could I just have a broken driver? 

![image](https://user-images.githubusercontent.com/83100346/206582249-f9657023-522c-495c-a3e4-6079ff63904d.png)

**Comment by cheat-engine** [comment]
yes. It looks like that whole block has been paged out to disk.  It's only available when it gets touched by the OS

There are infinite reasons why opening a driver fails.  Might be a faulty driver, or the driver only allows it to be opened by valid controllers for which there are many different ways it can check

**Comment by Nordgaren** [comment]
This is the Driver we are using. It's a training driver, I am sure you have heard of it. I did not compile the driver myself, but we have disabled 
<https://github.com/hacksysteam/HackSysExtremeVulnerableDriver>

Also, I can see the driver in WinDBG, I set nt!Kd_DEFAULT Mask and nt!Kd_IHVDRIVER_Mask and I see no output in windbg from HEVD, so it's probably a skill issue on my side.

So I am guessing no way to view this, or would I be able to catch it somehow with scripting after I figure out how to get my driver to load? 😆 

Thank you again for everything you do!

**Comment by Nordgaren** [comment]
Okay, so part of my skill issue was that python 3.8.11 is somehow not working for CreateFileA.

Now that I have gotten a drive handle, it is indeed no longer paged! I am currently trying CE 7.2, instead of 7.4.

**Comment by Nordgaren** [comment]
So, in 7.1, I can attach the debugger and set a break-point, but I get no break. in 7.4 I get this

![image](https://user-images.githubusercontent.com/83100346/206608839-e05f61ea-50bd-40f3-946d-96f9b35f9e0a.png)

Maybe I will try 7.3, but do you know what could have changed between these versions, and are the registry entries for Use Kernel Debugger in 7.1 missing a weird thing, or were they not there, in that version?
Thanks, again!

**Comment by Nordgaren** [comment]
last update for the night. I was able to get 7.0 to have the kernel mode debugger checked, even though it's still greyed out, but it still doesn't hit the break point :( 

I even added a part in the python script that gets the handle and triggers the IOCTL to pause before `DeviceIoControl` and then when and targeted python in the Cheat Engine, but still no breakpoint.

Hmm.

![image](https://user-images.githubusercontent.com/83100346/206610640-5ea742fb-27e0-44f4-bc8d-905548f0c8d4.png)
